processor/src/host/advice/mod.rs

```diff
@@ -0,0 +1,768 @@
+pub use inputs::AdviceInputs;
+
+mod injectors;
+
+mod providers;
+pub use providers::{MemAdviceProvider, RecAdviceProvider};
+
+mod source;
+pub use source::AdviceSource;
+
+// ADVICE PROVIDER
+// ================================================================================================
+
+/// Defines behavior of an advice provider.
+///
+/// An advice provider is a component through which the host can interact with the advice provider.
+/// The host can request nondeterministic inputs from the advice provider (i.e., result of a
+/// computation performed outside of the VM), as well as insert new data into the advice provider.
+///
+/// An advice provider consists of the following components:
+/// 1. Advice stack, which is a LIFO data structure. The processor can move the elements from the
+///    advice stack onto the operand stack, as well as push new elements onto the advice stack.
+/// 2. Advice map, which is a key-value map where keys are words (4 field elements) and values are
+///    vectors of field elements. The processor can push the values from the map onto the advice
+///    stack, as well as insert new values into the map.
+/// 3. Merkle store, which contains structured data reducible to Merkle paths. The VM can request
+///    Merkle paths from the store, as well as mutate it by updating or merging nodes contained in
+///    the store.
+pub trait AdviceProvider: Sized {
+    // ADVICE HANDLERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Handles the specified advice injector request.
+    fn set_advice<S: ProcessState>(
+        &mut self,
+        process: &S,
+        advice_injector: &AdviceInjector,
+    ) -> Result<HostResponse, ExecutionError> {
+        match advice_injector {
+            AdviceInjector::MerkleNodeMerge => self.merge_merkle_nodes(process),
+            AdviceInjector::MerkleNodeToStack => self.copy_merkle_node_to_adv_stack(process),
+            AdviceInjector::MapValueToStack {
+                include_len,
+                key_offset,
+            } => self.copy_map_value_to_adv_stack(process, *include_len, *key_offset),
+            AdviceInjector::UpdateMerkleNode => {
+                let depth = process.get_stack_item(4);
+                let index = process.get_stack_item(5);
+                let old_root = [
+                    process.get_stack_item(9),
+                    process.get_stack_item(8),
+                    process.get_stack_item(7),
+                    process.get_stack_item(6),
+                ];
+                let new_node = [
+                    process.get_stack_item(13),
+                    process.get_stack_item(12),
+                    process.get_stack_item(11),
+                    process.get_stack_item(10),
+                ];
+                let (path, _) = self.update_merkle_node(old_root, &depth, &index, new_node)?;
+                Ok(HostResponse::MerklePath(path))
+            }
```

Comment:
I wonder if there is a good way to make this more consistent with other injector handlers. Specifically, maybe have some intermediate method which gets the data from the process and then calls `update_merkle_node()`. One benefit of this is that if someone wants to implement `AdviceProvider` trait and override `set_advice()` method, they would not need to duplicate this logic.